### PR TITLE
Spot light cookies support by the Clustered lighting.

### PR DESCRIPTION
**Pull request by mvaligursky**
This implements part of https://github.com/playcanvas/engine/issues/3522 - cookie texture support for spot lights. This, similarly as clustered shadows, uses texture atlas to store cookie textures for visible lights, and fragment shader handling the light fetches appropriate cookie from it. This shares the same atlas layout as shadow atlas.

Noteworthy info:
- clustered cookies do not support cookie texture transform (scale and offset is not easy to do in an atlas .. postponing to later / when needed). (rotation can be done by light rotation, no need for texture rotation)
- clustered cookies do not support _Cookie Falloff_ property (this is doable, but not done unless needed)
- old lighting system supports any channel swizzle when specifying cookie texture. Editor only exposes common types: r, g, b, a, rgb. Only these common types are supported by clustered lighting (for example .rgg does not work)

Screenshot 2021-09-27 at 14 34 20

### CODE REVIEW

File: src/framework/components/light/component.js [hunk 1]
```diff
@@ -114,7 +114,7 @@ var _lightPropsDefault = [];
  * @property {Texture} cookie Projection texture. Must be 2D for spot and cubemap for omni light (ignored if incorrect type is used).
  * @property {number} cookieIntensity Projection texture intensity (default is 1).
  * @property {boolean} cookieFalloff Toggle normal spotlight falloff when projection texture is used. When set to false, spotlight will work like a pure texture projector (only fading with distance). Default is false.
- * @property {string} cookieChannel Color channels of the projection texture to use. Can be "r", "g", "b", "a", "rgb" or any swizzled combination.
+ * @property {string} cookieChannel Color channels of the projection texture to use. Can be "r", "g", "b", "a", "rgb".
```
Comment (willeastcott) [inline src/framework/components/light/component.js:117]: This is fine - I don't think anyone will be using a swizzle.

File: src/graphics/program-lib/chunks/clusteredLight.frag [hunk 2]
```diff
@@ -20,26 +22,58 @@ uniform vec3 clusterCellsMax;
...
+    // world space position
     vec3 position;
+
+    // world space direction (spot light only)
     vec3 direction;
+
+    // range of the light
     float range;
+
+    // spot light inner and outher angle cosine
```
Comment (willeastcott) [inline src/graphics/program-lib/chunks/clusteredLight.frag:59]:
```suggestion
    // spot light inner and outer angle cosine
```

File: src/scene/lighting/light-texture-atlas.js [hunk 3]
```diff
@@ -19,32 +23,69 @@ class LightTextureAtlas {
...
 
+    allocateCookieMap(resolution) {
+        if (!this.cookieMap || this.cookieMap.width !== resolution) {
+            this.destroyCookieMap();
+            this.cookieMap = CookieRenderer.createTexture(this.device, resolution);
+
+            this.cookieRenderTarget = new RenderTarget({
+                colorBuffer: this.cookieMap,
+                depth: false,
+                flipY: true // ///// ????
```
Comment (willeastcott) [inline src/scene/lighting/light-texture-atlas.js:79]: Interesting comment! 😆

**Review by willeastcott — APPROVED** [review]

File: src/scene/lighting/light-texture-atlas.js (hunk 3 above)
Comment (mvaligursky) [reply src/scene/lighting/light-texture-atlas.js]: hah yeah, visible to not miss it when creating PR. Didn't work. Fixed.

**Pull request merged**